# k-NN Classification

Sentiment classification on the IMDB Dataset using k-NN Classification to predict positive or negative reviews.

In [ ]:
import numpy as np
from tensorflow.keras.datasets import imdb
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

In [ ]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=10000)

def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, seq in enumerate(sequences):
        results[i, seq] = 1
    return results

x_train_vec = vectorize_sequences(x_train)
x_test_vec = vectorize_sequences(x_test)

In [ ]:
knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(x_train_vec, y_train)
y_pred_knn = knn_model.predict(x_test_vec)

In [ ]:
print("k-NN Accuracy:", accuracy_score(y_test, y_pred_knn))
print(classification_report(y_test, y_pred_knn))

k-NN Accuracy: 0.62068
              precision    recall  f1-score   support

           0       0.64      0.54      0.59     12500
           1       0.60      0.70      0.65     12500

    accuracy                           0.62     25000
   macro avg       0.62      0.62      0.62     25000
weighted avg       0.62      0.62      0.62     25000



The k-NN model achieves about 62.07% accuracy on the IMDB sentiment classification task. The precision, recall, and F1-scores around 0.62 indicate moderate performance, with noticeable difficulty in correctly classifying reviews across both classes due to the sparse and high-dimensional bag-of-words representation.